docs/models_dev_guide.md

```diff
@@ -0,0 +1,263 @@
+# Models Developer's Guide
+
+This document describes the design of the model management system in ETA and
+best practices for using it.
+
+
+## Motivation
+
+The ETA library relies on a number of large models in order to power its
+analytics. Models are trained on different machines, by different developers,
+and constantly need to be shared. We frequently generate new versions of a
+given model over time as we improve the implementation of the analytic and
+train it on evolving datasets. In addition, models are valuable, so they need
+to be archived and backed up to ensure that they remain secure and readily
+accessible.
+
+The ETA library provides a simple, yet powerful interface that satisifies
+these constraints.
+
+
+## Design
+
+The `eta.core.models` module implements the core interface for managing models
+in ETA. The underlying design principle of the interface is that all models
+are stored in cloud storage and can be easily downloaded to a local ETA
+installation as needed using a simple version-aware naming system.
+
+### Model versions
+
+Each model is identified by a unique string of the form `<name>@<version>`,
```

Comment:
How is uniqueness of strings enforced across developers?

Reply:
Uniqueness is forced via the `manifest.json` files on your models search path. Basically any models-related operation you perform will cause all manifests to be read. If there is a name conflict at load time, an error will be raised.

Note that we need to commit the manifest files to the repo, because this is how we "expose" our models across the team.

We can also have our own "private" models for each developer by simply adding our own private folders to our search paths and not adding the relevant `manifest.json` to version control.